aoservicelisting.civix.php

```diff
@@ -173,36 +193,142 @@ public static function endABARelationship($values, $rowNumber, $orgId) {
+          $currentDetails[REGULATED_URL] != $individualParams['regulated_url']) {
+          // This is an overwrite on the record, check to see if this contact has an ABA certificate before ending the relationship.
+          if (!empty($currentDetails[CERTIFICATE_NUMBER])) {
+            // Move the certificate number elsewhere and delete the certificate number, don't end the relationship.
+            E::archiveField(CERTIFICATE_NUMBER, $currentDetails[CERTIFICATE_NUMBER], $cid);
+          }
+          else {
+            // There is no certificate number, we can now end the relationship.
+            E::terminateRelationship($cid, $orgId);
+          }
+        }
+        else {
+          // This is an edit, return the contact ID.
+          if ($currentDetails['first_name'] == $individualParams['first_name'] &&
+            $currentDetails['last_name'] == $individualParams['last_name'] &&
+            $currentDetails[REGULATED_URL] != $individualParams['regulated_url']) {
+            // This is an edit to only the regulated URL field, primary contacts can be edited.
+            $individualParams['contact_id'] = $cid;
+            return;
+          }
+          if (($currentDetails['first_name'] != $individualParams['first_name'] ||
+            $currentDetails['last_name'] != $individualParams['last_name']) &&
+            $currentDetails[REGULATED_URL] == $individualParams['regulated_url']) {
+            // This is an edit to the name, ensure that it is not a primary contact.
+            if (E::checkPrimaryContact($cid)) {
+              return;
+            }
+            else {
+              $individualParams['contact_id'] = $cid;
+            }
+          }
+          if ($currentDetails['first_name'] == $individualParams['first_name'] &&
+            $currentDetails['last_name'] == $individualParams['last_name'] &&
+            $currentDetails[REGULATED_URL] == $individualParams['regulated_url']) {
+            // No edits, return the contact id.
+            $individualParams['contact_id'] = $cid;
+          }
+        }
+      }
+      elseif ($staffType == 'abastaff') {
+        // Get saved certificate number
+        if ($currentDetails['first_name'] != $individualParams['first_name'] &&
+          $currentDetails['last_name'] != $individualParams['last_name'] &&
+          $individualParams[CERTIFICATE_NUMBER] != $currentDetails[CERTIFICATE_NUMBER]) {
+          // This is an overwrite on the record, check to see if this contact has a regulated URL before ending the relationship.
+          if (!empty($currentDetails[REGULATED_URL])) {
+            // Move the certificate number elsewhere and delete the certificate number, don't end the relationship.
+            E::archiveField(REGULATED_URL, $currentDetails[REGULATED_URL], $cid);
+          }
+          else {
+            // There is no certificate number, we can now end the relationship.
+            E::terminateRelationship($cid, $orgId);
```

Comment:
don't we need to check for pirmary contact here and not terminate relationship if primary? and archive ABA field if primary?